fix(compiler-cli): more accurate diagnostics for host binding parser errors

crisbeto opened the pull request:
Currently host bindings are in a bit of a weird state, because their source spans all point to the root object literal, rather than the individual expression. This is tricky to handle at the moment, because the object is being passed around as a `Record<string, string>` since the compiler needs to support both JIT and non-JIT environments, and because the AOT compiler evaluates the entire literal rather than doing it expression-by-expression. As a result, when we report errors in one of the host bindings, we end up highlighting the entire expression which can be very noisy in an IDE.

These changes aim to report a more accurate error for the most common case where the `host` object is initialized to a `string -> string` object literal by matching the failing expression to one of the property initializers. Note that this isn't 100% reliable, because we can't map cases like `host: SOME_CONST`, but it's still better than the current setup.

**Before:**
<img width="927" alt="image" src="https://github.com/user-attachments/assets/7d1a0312-5d42-48c3-b0af-a938d57b6d6e">


**After:**
<img width="927" alt="image" src="https://github.com/user-attachments/assets/526b0c4d-11b9-4ba0-924a-73a016f22953">

JoostK reviewed (approved):
One additional gotcha here could be for duplicate host binding expressions; they would report individual errors for each occurrence, but will then all get reported on the first host binding occurrence.


Overall this still seems like an improvement :+1:

crisbeto commented:
I was thinking about duplicates as well, but I don't think they'll be a problem because we throw a `FatalDiagnosticError` for the first invalid expression we see so the rest won't be reported.

crisbeto committed:
fix(compiler-cli): more accurate diagnostics for host binding parser errors

Currently host bindings are in a bit of a weird state, because their source spans all point to the root object literal, rather than the individual expression. This is tricky to handle at the moment, because the object is being passed around as a `Record<string, string>` since the compiler needs to support both JIT and non-JIT environments, and because the AOT compiler evaluates the entire literal rather than doing it expression-by-expression. As a result, when we report errors in one of the host bindings, we end up highlighting the entire expression which can be very noisy in an IDE.

These changes aim to report a more accurate error for the most common case where the `host` object is initialized to a `string -> string` object literal by matching the failing expression to one of the property initializers. Note that this isn't 100% reliable, because we can't map cases like `host: SOME_CONST`, but it's still better than the current setup.
